### PR TITLE
Edge - JSON converter for proto

### PR DESCRIPTION
**Pull request by AndriiLandiak**
## Pull Request description

**Issue**:
Currently, maintaining backward compatibility with Edge when new fields are added to entities is cumbersome. It involves updating proto configurations and extensive testing after any Tb entity updates.

**Solution**:
To simplify this process, this PR uses serialization and deserialization for sending entities. By using Object Mapper serialization and setting FAIL_ON_UNKNOWN_PROPERTIES to **false** during deserialization, we can seamlessly handle different entity versions between our Edge and TB.

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [x] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.
- [x] If new yml property was added: make sure a description is added (above or near the property).

### CODE REVIEW

**Review by volodymyr-babak — APPROVED** [review]

**Comment by volodymyr-babak** [comment]
Edge processors versioning is planned as well, but as a separate pull request.

File: common/edge-api/src/main/proto/edge.proto [hunk 1]
```diff
@@ -138,25 +139,28 @@ message RuleChainUpdateMsg {
...
-  bool root = 7;
-  bool debugMode = 8;
-  string configuration = 9;
+  string name = 4 [deprecated = true];
+  optional int64 firstRuleNodeIdMSB = 5 [deprecated = true];
+  optional int64 firstRuleNodeIdLSB = 6 [deprecated = true];
+  bool root = 7 [deprecated = true];
+  bool debugMode = 8 [deprecated = true];
+  string configuration = 9 [deprecated = true];
+  string entity = 10;
```
Comment (smatvienko-tb) [inline common/edge-api/src/main/proto/edge.proto:148]: I am on a similar topic where a Proto version is needed for different versions.

For your case, the edge.proto is a speaking language between endpoints. The endpoints may have different versions, but the language (proto) should be **the same** or at least **fully compatible**.

For example, you add a mandatory field `string entity = 10;` that is not optional, so it is not compatible with the old proto version.

I assume that you have no fully integrated test of the ThingsBoard/master  + Edges with all compatible versions. May be I wrong.

As far as I see, you are trying to maintain compatibility with many older versions and the way to do this solid is to retain all the proto versions you need. When you establish the gRPC session, you can negotiate the proto version and than proceed with a compatible serializer/deserializer.

One of the ways to do this is to use the version number in the package.

For now, you are on v1 despite making major changes to the speaking language.


- Your messaging is not persistent, and it is much simple to establish the compatible proto version and process it accordingly.

On the next major version, you can cut off some compatibility and remove unsupported proto from the project



```
option java_package = "org.thingsboard.server.gen.edge.v1";
```

**Review by smatvienko-tb — COMMENTED** [review]
I've included my comment on the proto-versioning strategy. 
I hope it will help you to make a rock-stable compatible release.

Here is some good reading on gRPC to my taste  https://learn.microsoft.com/en-us/aspnet/core/grpc/versioning?view=aspnetcore-8.0

File: common/edge-api/src/main/proto/edge.proto (hunk 1 above)
Comment (AndriiLandiak) [reply common/edge-api/src/main/proto/edge.proto:148]: Thank you for your insights. We will carefully consider the most appropriate approach for handling deprecated fields, and it seems that opting for a major version cut-off might be the most suitable solution.

According to compatibility with the previous Edge, I should admit that backward compatibility works fine when introducing new fields without the 'optional' keyword, as exemplified by our usage of string entity = 10. However, it's worth noting that the Protocol Buffers documentation recommends adding new fields as either 'optional' or 'repeated'.

_`Make all fields optional or repeated. You never know how long a message type is going to last and whether someone will be forced to fill in your required field with an empty string or zero in four years when it’s no longer logically required but the proto still says it is.`_  [Proto best practices](https://protobuf.dev/programming-guides/dos-donts/)

Comment (AndriiLandiak) [reply common/edge-api/src/main/proto/edge.proto:148]: Thank you for your insights. We will carefully consider the most appropriate approach for handling deprecated fields, and it seems that opting for a major version cut-off might be the most suitable solution.

> For example, you add a mandatory field string entity = 10; that is not optional, so it is not compatible with the old proto version.

According to compatibility with the previous Edge, I should admit that backward compatibility works fine when introducing new fields without the 'optional' keyword, as it was by our usage of string entity = 10. However, it's worth noting that the Protocol Buffers documentation recommends adding new fields as either 'optional' or 'repeated'.

Make all fields optional or repeated. You never know how long a message type is going to last and whether someone will be forced to fill in your required field with an empty string or zero in four years when it’s no longer logically required but the proto still says it is. [Proto best practices](https://protobuf.dev/programming-guides/dos-donts/)

Comment (ashvayka) [reply common/edge-api/src/main/proto/edge.proto:148]: As a suggestion:

```
message EntityUpdateMsg {
  UpdateMsgType msgType = 1;
  string entityType = 2
  int64 idMSB = 3;
  int64 idLSB = 4;
  string entity = 5;
}
```

**Pull request merged**